### PR TITLE
N-ary multiplication

### PR DESCRIPTION
**Pull request by mikand**
As explained in issue #297, this PR transforms the Times operator in a n-ary node.

All the components of the system have been ported to this change.

### CODE REVIEW

File: pysmt/oracles.py [hunk 1]
```diff
@@ -223,14 +223,12 @@ def walk_lira(self, formula, args, **kwargs):
...
         theory_out = args[0]
         for t in args[1:]:
             theory_out = theory_out.combine(t)
-        # Check for non-linear
-        left, right = formula.args()
-        if len(left.get_free_variables()) != 0 and \
-           len(right.get_free_variables()) != 0:
+        # Check for non-linear counting the arguments having at least
+        # one free variable
+        if sum(1 for x in formula.args() if len(x.get_free_variables()) != 0) > 1:
```
Comment (marcogario) [inline pysmt/oracles.py]: ```if any(len(x.get_free_variables()) != 0):```  
(you can avoid the len by simply checking ```if x.get_free_variables()```).

File: pysmt/simplifier.py [hunk 2]
```diff
@@ -288,44 +288,37 @@ def walk_plus(self, formula, args, **kwargs):
...
-                return sl
-            elif sr.is_zero():
-                if sr.is_real_constant():
-                    return self.manager.Real(0)
-                else:
-                    assert sr.is_int_constant()
-                    return self.manager.Int(0)
+                return self.manager._Algebraic(Numeral(mul))
+
+        new_args = []
```
Comment (marcogario) [inline pysmt/simplifier.py]: I think we should unify both cases. In particular, we iterate on the arguments and use an accumulator for the constants, while maintaining the non-constant objects. Currently you are doing 2 passes of the arguments. I think you can do it in one.

* If it is zero: return 0
* If it is one: continue
* If is is a constant: add to accumulator
* If is is not a constant, add to new_args

Do you have anther reason for this?

**Comment by marcogario** [comment]
:+1: Ok to merge once you handle the two comments above.

Note: I added one example with 0 in the multiplication, just to be sure that it is handled correctly.

File: pysmt/solvers/cvc4.py [hunk 3]
```diff
@@ -288,9 +288,16 @@ def walk_equals(self, formula, args, **kwargs):
         return self.mkExpr(CVC4.EQUAL, args[0], args[1])
 
     def walk_times(self, formula, args, **kwargs):
-        if not args[0].isConst() and not args[1].isConst():
-            raise NonLinearError(formula)
-        return self.mkExpr(CVC4.MULT, args[0], args[1])
+        res = args[0]
+        nl_count = 0 if res.isConst() else 1
```
Comment (marcogario) [inline pysmt/solvers/cvc4.py]: Looks like isConst() returns whether you have a constant term and not a constant expression.

https://travis-ci.org/pysmt/pysmt/jobs/153900213

**Comment by marcogario** [comment]
This seems to fail also in the SMT-LIB wrappers. Maybe times is only binary in SMT-LIB?

File: pysmt/oracles.py (hunk 1 above)
Comment (mikand) [inline pysmt/oracles.py]: ```if any(len(x.get_free_variables()) != 0):```
is not correct.

Here we need to check if at least two arguments are non-constant (hence I check if the count is > 1). Surely we can avoid the len, but I feel that leaving it is clearer.

File: pysmt/simplifier.py (hunk 2 above)
Comment (mikand) [inline pysmt/simplifier.py]: :+1:

**Pull request merged**